//option optimize_for = LITE_RUNTIME;
package user;
import public "storage.proto";
import public "stats.proto";

enum MessageType{
	reqLogin = 1; 		// N/A
//	reqLogout; 		// N/A 
//	reqItem; 		// N/A
//	reqParameters; 	// N/A
//	reqCases; 		// N/A
//	
//	addUser;		// N/A
//	addParameter;	// N/A
//	addItemToStorage;// N/A
//	addItemGroup;	// N/A
//	addItem;		// N/A
//	addCase;		// N/A
//	addFileToUser;	// N/A
//	addFileToGroup;	// N/A
//	addFileToItem;	// N/A
//	addFileToCase;	// N/A
//	addStorage;		// N/A
//	addSet;			// N/A
}

enum Replay{
  LOGIN_OK = 1;
  BAD_USER_OR_PASSWD = 10;
  USER_ALREADY_LOGGED_IN = 11;
}

message LoginRequest{
  required string name = 3;
  required string password = 4;
}

message LoginResponse{
  required uint32 msgType = 1[default = 14];
  required Replay replay = 2;
}

message Add{
  required string name = 2;
  required string password = 3;
  required string email = 4;
  optional string address = 5;
  optional bytes phonenumber = 8;
}

message AddReplay{
  required QueryStats stats = 2;
  required Replay replay = 3;
}

message UserBasicData{
    required string name = 3;
    required string email = 4;
}

message UserData{
    required uint32 ID = 2;
    required string name = 3;
    required string email = 4;
    repeated storage.Storage storages = 11;
    optional string address = 5;
    optional bytes phoneNumber = 6;
    optional uint64 lastLogin = 7; // N/A
    optional uint64 firstLogin = 8; // N/A
    optional bytes userConfig = 9; // windows size, program settings, last group, etc.
    optional UserActivityStatistics stats = 10; // N/A
}

message UserActivityStatistics{
	required uint32 magazineCount = 1; // number of magazines
	required uint32 totalNumOfElements=2; // number of elements
	required uint32 totalNumOfUniqueElements = 3; // number of different elements
	required uint32 totalNumOfRequests = 5; // number of requests to server
	required uint32 totalModifiedItems = 6; // total number of modified elements by user
	required uint32 totalAddItems = 7; // total items that user add to groups
}
